### PR TITLE
Add UPath dependency/type checking

### PR DESCRIPTION
**Pull request by dstansby**
Fixes some kwarg typing, and adds `universal-pathlib` to the type checking and as a package dependency.

### CODE REVIEW

**Review by jhamman — APPROVED** [review]

File: pyproject.toml [hunk 1]
```diff
@@ -29,6 +29,7 @@ dependencies = [
     'numcodecs>=0.10.0',
     'fsspec>2024',
     'crc32c',
+    'universal-pathlib',
```
Comment (normanrz) [inline pyproject.toml:32]: Should we make upath an optional dependency?

Comment (tomwhite) [reply pyproject.toml:32]: Yes please!

Comment (dstansby) [reply pyproject.toml]: 👍 done

File: src/zarr/store/remote.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 from typing import TYPE_CHECKING, Any
 
 import fsspec
+from upath import UPath
```
Comment (normanrz) [inline src/zarr/store/remote.py:7]: This makes it a required dependency, no?

**Review by jhamman — APPROVED** [review]

**Comment by dstansby** [comment]
pre-commit-ci autofix

**Pull request merged**